### ISSUE TITLE
Test console output capturing/swallowing

### ISSUE THREAD

**Issue by roji** [xunit/xunit#2071]
When running tests, Console.Out sometimes shows and sometimes does not. I'm well aware that XUnit 2.0 stopped capturing Console.Out, and that XUnit isn't supposed to be interfering with output in any way (e.g. [this comment seems to point to VSTest](https://github.com/xunit/xunit/issues/1141#issuecomment-555745227)).

However, I've done a quick comparison between XUnit and NUnit on this question and have come up with the following:

|                                            | XUnit | NUnit |
|--------------------------------------------|-------|-------|
| dotnet test (SDK 3.1.101, Windows)         | No    | Yes\*    |
| dotnet test (SDK 3.1.101, Linux)           | Yes   | Yes\*    |
| JetBrains Rider (2019.3.3 Linux & Windows) | No    | Yes   |
| VS 16.4.5 + Resharper 2019.3.3             | No    | Yes   |
| VS 16.4.5 (native test runner)             | No    | Yes   |

\* Requires NUnit3TestAdapter 3.17.0-dev-01324 and `dotnet test -v n`

In other words, Console.WriteLine shows up just fine under NUnit in all IDEs; this at least suggests that the IDEs/VSTest do support Console.Out and that something being done in XUnit (or xunit.runner.visualstudio?) could be responsible. The dotnet test situation is confusing - I'm seeing different results on Windows and Linux. I'm not blaming XUnit in any way - just sharing information, and hoping for us to understand the situation better. If it does end up being a VSTest issue, I'd be happy to try to push that forward with the relevant people.

The repo used to test XUnit vs. NUnit is https://github.com/roji/TestConsoleTests.

Note: issue tracking Console.WriteLine under dotnet test for NUnit: https://github.com/nunit/nunit3-vs-adapter/issues/343

**Comment by roji** [comment]
Note: with recent prerelease versions of NUnit, Console.WriteLine can be seen with dotnet test as well, making NUnit work in all cases.

**Comment by gandarez** [comment]
I'm also having the same problem using xUnit.

**Comment by roji** [comment]
Relevant issue, including on the Windows/Linux difference: https://github.com/microsoft/vstest/issues/799

**Comment by MatteS75** [comment]
This is the biggest issue for us with xunit.

**Comment by SimonCropp** [comment]
not a fix, but a temporary workaround that redirects Trace.Write, Debug.Write, and Console.Write and Console.Error.Write to ITestOutputHelper. https://github.com/SimonCropp/XunitContext

**Comment by moh-hassan** [comment]
In Windows, Running `dotnete test` from within Powershell Console like Powershell ISE or Package Console Manager in VS (tested in vs 2019) capture all Console.WriteLine used inside Xunit project or in source code. 
here script:

cls
cd "path\to\test\project"
dotnet test

**Comment by krwq** [comment]
I'm suspecting this is this issue https://github.com/microsoft/vstest/issues/2224

**Comment by naavis** [comment]
I've been bumping into this problem lately too. Any updates?

If I create a dummy xUnit project in Visual Studio and make one test that just writes to console with `Console.WriteLine("Hello World");`, and run it with `dotnet test`, the visibility of that console message seems to depend on many things.

- In normal Powershell 5.1 (in Windows Terminal on Windows 10) I don't see the message. I am however able to detect it by piping output to a file with `dotnet test > foo.log`. Trying to redirect all output to the Success stream of Powershell with `dotnet test *>&1` does not show the message, however, which is very curious.
- In Powershell ISE's integrated console I see the message without doing anything.
- In normal cmd.exe command prompt I do not see the console output either.
- Also running the test in an Azure DevOps pipeline I see the console message in the output of the test job.

Trying the same with NUnit instead of xUnit I never see the console output.

**Comment by bradwilson** [comment]
We will offer an optional ability to capture stdout/stderr/Trace/Debug in v3, but there are no plans to change this in v2.

**Comment by naavis** [comment]
I'm more curious about the root cause of the output being captured in some cases and not being captured in others, and is there anything I can do about it as a user of xUnit. I don't mind not being able to see the console output of tests, nor always seeing it, but inconsistently showing it in some cases but not others is weird.

**Comment by SimonCropp** [comment]
@naavis out of curiosity, does this meet your requirements https://github.com/SimonCropp/XunitContext ?

**Comment by bradwilson** [comment]
@naavis The inconsistency comes from the fact that there are many runners, and we don't control them all. The behavior that VSTest (which powers Test Explorer and `dotnet test`) chooses to implement is out of our control.

**Comment by jcracknell** [comment]
This actually turns out to be fairly easy to implement. You create a central (static) registry which maintains a list of active ITestOutputHelper instances as well as setting an AsyncLocal instance. You create a write method which either (a) writes to the AsyncLocal instance, as in this case the output was caused directly by the associated test case, or (b) writes to all registered instances, as the output was triggered by a background task (i.e. your test server) and we don't know the associated task.

Override XunitTestRunner

**Comment by peabnuts123** [comment]
I believe this will be / has been resolved by microsoft/vstest#4998
